File: claude_config/.claude/hooks/validate_html.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.8"
# dependencies = [
#     "beautifulsoup4",
# ]
# ///
"""
Validation script for HTML restructuring output.

This script validates that:
1. The output HTML file contains valid HTML
2. All source IDs from input are covered by data-sources attributes in output
3. No extra IDs are present in output that weren't in input

Usage: uv run --script validate_html.py [--block] <input_html_file> <output_html_file>

Options:
    --block    Use blocking error codes (exit 2) for failures instead of non-blocking (exit 3)
"""

import re
import sys
import argparse
import bs4
from pathlib import Path

ALLOWED_TAGS = [
    "header", "main", "footer", "figure", "figcaption",
    "table", "thead", "tbody", "tfoot", "th", "tr", "td", "caption",
    "section", "nav", "aside", "p", "ul", "ol", "li", "h1",
    "h2", "h3", "h4", "h5", "h6", "img", "math", "code",
    "cite", "blockquote", "b", "i", "u", "s", "sup", "sub", "br"
]

# Inline style tags that don't affect leaf node status
INLINE_STYLE_TAGS = {"b", "i", "u", "s", "sup", "sub", "code", "cite", "br"}

def parse_range_string(range_str: str) -> list[int]:
    """
    Parse a comma-separated range string into a list of integers.
    
    Args:
        range_str: String like "0-3,5,7-9" or empty string
        
    Returns:
        List of integers representing all tag ids in the ranges
        
    Raises:
        ValueError: If the range string format is invalid
    """
    if not range_str.strip():
        return []
    
    ids: list[int] = []
    parts = range_str.split(",")
    
    for part in parts:
        part = part.strip()
        if "-" in part:
            # Handle range like "1-3"
            range_parts = part.split("-", 1)
            if len(range_parts) != 2:
                raise ValueError(f"Invalid range format: {part}")
            try:
                start_num = int(range_parts[0].strip())
                end_num = int(range_parts[1].strip())
                if start_num > end_num:
                    raise ValueError(f"Invalid range: start ({start_num}) > end ({end_num})")
                ids.extend(range(start_num, end_num + 1))
            except ValueError as e:
                if "invalid literal" in str(e):
                    raise ValueError(f"Non-numeric values in range: {part}")
                raise
        else:
            # Handle single number like "5"
            try:
                ids.append(int(part))
            except ValueError:
                raise ValueError(f"Non-numeric value: {part}")
    
    return sorted(ids)


def validate_html_structure(input_file: Path, output_file: Path) -> tuple[bool, str]:
    """
    Validate that output HTML properly covers all input IDs.
    
    Returns:
        Tuple of (is_valid, error_message)
    """
    try:
        # Read input HTML
        if not input_file.exists():
            return False, f"Input file does not exist: {input_file}"
        
        with open(input_file, "r", encoding="utf-8") as f:
            input_html = f.read()
        
        # Read output HTML
        if not output_file.exists():
            return False, f"Output file does not exist: {output_file}"
        
        with open(output_file, "r", encoding="utf-8") as f:
            output_html = f.read()
        
        # Replace <em> with <i> and <strong> with <b> in output
        replacements_made = []
        if "<em>" in output_html or "</em>" in output_html:
            output_html = output_html.replace("<em>", "<i>").replace("</em>", "</i>")
            replacements_made.append("<em> → <i>")
        if "<strong>" in output_html or "</strong>" in output_html:
            output_html = output_html.replace("<strong>", "<b>").replace("</strong>", "</b>")
            replacements_made.append("<strong> → <b>")
        
        # Write back the modified output if replacements were made
        if replacements_made:
            with open(output_file, "w", encoding="utf-8") as f:
                f.write(output_html)
            print(f"📝 Auto-replaced tags: {', '.join(replacements_made)}", file=sys.stderr)
        
        # Parse HTML
        try:
            input_soup = bs4.BeautifulSoup(input_html, "html.parser")
            output_soup = bs4.BeautifulSoup(output_html, "html.parser")
        except Exception as e:
            return False, f"Failed to parse HTML: {e}"
        
        # Extract IDs from input
        ids_in_input: set[int] = set()
        for element in input_soup.find_all():
            if "id" in element.attrs:
                try:
                    ids_in_input.add(int(element.attrs["id"]))
                except ValueError:
                    return False, f"Non-numeric ID found in input: {element.attrs['id']}"
        
        # Check for disallowed tags in output (only within body)
        disallowed_tags: set[str] = set()
        body = output_soup.find("body")
        if body:
            for element in body.find_all():
                if element.name and element.name not in ALLOWED_TAGS:
                    disallowed_tags.add(element.name)
        else:
            # If no body, check all elements
            for element in output_soup.find_all():
                if element.name and element.name not in ALLOWED_TAGS:
                    disallowed_tags.add(element.name)
        
        # Extract IDs from output data-sources attributes (leaf nodes only)
        ids_in_output: set[int] = set()
        for element in output_soup.find_all():
            if "data-sources" in element.attrs:
                # Check if this is a leaf node (has no child elements with tags, excluding inline style tags)
                has_child_elements = any(
                    child.name and child.name not in INLINE_STYLE_TAGS 
                    for child in element.children 
                    if hasattr(child, 'name')
                )
                if not has_child_elements:
                    try:
                        ids_in_output.update(parse_range_string(element["data-sources"]))
                    except Exception as e:
                        return False, f"Failed to parse data-sources '{element['data-sources']}': {e}"
        
        # Check coverage
        missing_ids = ids_in_input - ids_in_output
        extra_ids = ids_in_output - ids_in_input
        
        if disallowed_tags or missing_ids or extra_ids:
            error_msg = []
            if disallowed_tags:
                error_msg.append(
                    f"You've used some HTML tags in the {output_file} file that are not allowed: {sorted(disallowed_tags)}.\n"
                    f"The allowed tags are: {', '.join(ALLOWED_TAGS)}.\n"
                    "Fix these tags and keep going! You're doing great!"
                )
            if missing_ids:
                missing_msg = f"IDs in the {input_file} file not yet covered by leaf nodes in the {output_file} file: {sorted(missing_ids)}. You'll need to add more leaf nodes (or make sure all existing leaf nodes have data-sources) that cover these ids."
                if len(missing_ids) < 30:
                    missing_msg += (
                        "\n\nNote: if any input nodes are empty or contain garbage characters "
                        "you think shouldn't be in the final output, you may attach their ids "
                        "as data-sources to neighboring output nodes to pass the id validation "
                        "check. Please do this sparingly."
                    )
                error_msg.append(missing_msg)
            if extra_ids:
                # Provide clarifying guidance when output references non-existent IDs
                error_msg.append(
                    "Output references IDs not present in input (invalid data-sources). "
                    f"Extra IDs: {sorted(extra_ids)}. "
                    "Ensure data-sources only contain IDs that exist in the input and split ranges "
                    "to avoid spanning missing IDs."
                )
            return False, "; ".join(error_msg)
        
        return True, "All IDs properly covered"
        
    except Exception as e:
        return False, f"Validation error: {e}"


def main():
    parser = argparse.ArgumentParser(
        description="Validate HTML restructuring output",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
This script validates that:
1. The output HTML file contains valid HTML
2. All source IDs from input are covered by data-sources attributes in output
3. No extra IDs are present in output that weren't in input
        """
    )
    parser.add_argument("input_file", type=Path, help="Input HTML file path")
    parser.add_argument("output_file", type=Path, help="Output HTML file path")
    parser.add_argument(
        "--block", 
        action="store_true", 
        help="Use blocking error codes (exit 2) for failures instead of non-blocking (exit 3)"
    )

    args = parser.parse_args()

    is_valid, message = validate_html_structure(args.input_file, args.output_file)

    if is_valid:
        # Read the output file to check for placeholders
        with open(args.output_file, 'r', encoding='utf-8') as f:
            output_html = f.read()

        possible_placeholders = []
        for line_num, line in enumerate(output_html.splitlines(), 1):
            if re.search(r'placeholder|<!--|\.\.\.', line):
                possible_placeholders.append(f"Line {line_num}: {line.strip()}")
        if possible_placeholders:
            print(f"🔍 All ids from the {args.input_file} file are present as data-sources in the {args.output_file} file and all tags are valid.\n")
            print("This *might* mean you're done, **BUT**! Please review the following lines to make sure you haven't hacked the validation check by using placeholders or truncated content.")
            print(f"🔍 Possible placeholders found in the {args.output_file} file:\n" + "\n".join(possible_placeholders), file=sys.stderr)

        with open(args.input_file, 'r', encoding='utf-8') as f:
            input_html = f.read()

        # If output file is <85% the length of the input file, print a warning
        if len(output_html) < 0.85 * len(input_html):
            print("🔍 The output file passed a validation check showing all input ids are present as data-sources, but it's suspiciously short compared to the input file, so you've probably cheated by using placeholders or truncated content.")
            print("Please review the output file to make sure it actually contains all the input content.")

        print(
            f"✅ All ids from the {args.input_file} file are present as data-sources in the {args.output_file} file and all tags are valid.\n"
            "This doesn't necessarily mean you're done, but it's a good sign.\n"
            "Once you've checked that the output HTML is well structured with semantic tags and all meaningful content is well-represented "
            f"by leaf nodes in the {args.output_file} file, you can mark your task complete.", file=sys.stderr)
        # Use non-blocking exit code so the success message is visible to Claude
        sys.exit(3)
    else:
        print(f"We're making progress! 😊 {message}", file=sys.stderr)
        # Use blocking or non-blocking exit code based on --block flag
        exit_code = 2 if args.block else 3
        sys.exit(exit_code)


if __name__ == "__main__":
    main()

File: opencode_config/.opencode/plugin/validate_html.ts
import type { Plugin } from "@opencode-ai/plugin"
import { readFile, writeFile } from "fs/promises"
import { existsSync } from "fs"

const ALLOWED_TAGS = [
  "header", "main", "footer", "figure", "figcaption",
  "table", "thead", "tbody", "tfoot", "th", "tr", "td", "caption",
  "section", "nav", "aside", "p", "ul", "ol", "li", "h1",
  "h2", "h3", "h4", "h5", "h6", "img", "math", "code",
  "cite", "blockquote", "b", "i", "u", "s", "sup", "sub", "br"
]

// Inline style tags that don't affect leaf node status
const INLINE_STYLE_TAGS = new Set(["b", "i", "u", "s", "sup", "sub", "code", "cite", "br"])

/**
 * Parse a comma-separated range string into a list of integers.
 * 
 * @param rangeStr String like "0-3,5,7-9" or empty string
 * @returns Array of integers representing all tag ids in the ranges
 * @throws Error if the range string format is invalid
 */
function parseRangeString(rangeStr: string): number[] {
  if (!rangeStr.trim()) {
    return []
  }
  
  const ids: number[] = []
  const parts = rangeStr.split(",")
  
  for (const part of parts) {
    const trimmedPart = part.trim()
    if (trimmedPart.includes("-")) {
      // Handle range like "1-3"
      const rangeParts = trimmedPart.split("-", 2)
      if (rangeParts.length !== 2) {
        throw new Error(`Invalid range format: ${trimmedPart}`)
      }
      
      const startNum = parseInt(rangeParts[0].trim())
      const endNum = parseInt(rangeParts[1].trim())
      
      if (isNaN(startNum) || isNaN(endNum)) {
        throw new Error(`Non-numeric values in range: ${trimmedPart}`)
      }
      
      if (startNum > endNum) {
        throw new Error(`Invalid range: start (${startNum}) > end (${endNum})`)
      }
      
      for (let i = startNum; i <= endNum; i++) {
        ids.push(i)
      }
    } else {
      // Handle single number like "5"
      const num = parseInt(trimmedPart)
      if (isNaN(num)) {
        throw new Error(`Non-numeric value: ${trimmedPart}`)
      }
      ids.push(num)
    }
  }
  
  return ids.sort((a, b) => a - b)
}

/**
 * Simple HTML parser to extract elements and their attributes
 */
function parseHTML(html: string) {
  const elements: Array<{
    name: string
    attrs: Record<string, string>
    isLeafNode: boolean
  }> = []
  
  // Find all opening tags (both self-closing and regular)
  const openingTagRegex = /<(\w+)([^>]*?)(?:\s*\/>)/g
  let match
  
  while ((match = openingTagRegex.exec(html)) !== null) {
    const [fullMatch, tagName, attrsStr] = match
    
    // Parse attributes (including hyphenated ones like data-sources)
    const attrs: Record<string, string> = {}
    const attrRegex = /([\w-]+)=["']([^"']*)["']/g
    let attrMatch
    while ((attrMatch = attrRegex.exec(attrsStr)) !== null) {
      attrs[attrMatch[1]] = attrMatch[2]
    }
    
    // For simplicity, assume non-void elements could be leaf nodes
    // This is a simplified check - in real implementation we'd need better parsing
    const voidElements = ['img', 'br', 'hr', 'meta', 'link', 'area', 'base', 'col', 'embed', 'input', 'source', 'track', 'wbr']
    const isVoid = voidElements.includes(tagName.toLowerCase())
    
    elements.push({
      name: tagName,
      attrs,
      isLeafNode: !isVoid // Simplified assumption
    })
  }
  
  return elements
}

/**
 * Lightweight HTML tree parser sufficient for validation needs
 */
type ElementNode = {
  name: string
  attrs: Record<string, string>
  children: ElementNode[]
  parent?: ElementNode
}

function parseHTMLTree(html: string): { root: ElementNode, allNodes: ElementNode[], bodyNode?: ElementNode } {
  const root: ElementNode = { name: "__root__", attrs: {}, children: [] }
  const allNodes: ElementNode[] = []

  const tagRegex = /<\/?([a-zA-Z][\w:-]*)([^>]*?)\/?\s*>/g
  const attrRegex = /([\w-]+)=["']([^"']*)["']/g
  const voidElements = new Set(['img', 'br', 'hr', 'meta', 'link', 'area', 'base', 'col', 'embed', 'input', 'source', 'track', 'wbr'])

  const stack: ElementNode[] = [root]
  let matchTree: RegExpExecArray | null

  while ((matchTree = tagRegex.exec(html)) !== null) {
    const full = matchTree[0]
    const rawName = matchTree[1] || ""
    const tagName = rawName.toLowerCase()
    const attrsStr = matchTree[2] || ""
    const isClosing = full.startsWith("</")
    const isSelfClosing = full.endsWith("/>") || voidElements.has(tagName)

    if (isClosing) {
      // Pop until matching tag or root
      for (let i = stack.length - 1; i > 0; i--) {
        if (stack[i].name === tagName) {
          stack.length = i // pop matched element
          break
        }
      }
      continue
    }

    // Opening or self-closing tag
    const attrs: Record<string, string> = {}
    let am: RegExpExecArray | null
    while ((am = attrRegex.exec(attrsStr)) !== null) {
      attrs[am[1]] = am[2]
    }

    const node: ElementNode = { name: tagName, attrs, children: [], parent: stack[stack.length - 1] }
    stack[stack.length - 1].children.push(node)
    allNodes.push(node)

    if (!isSelfClosing) {
      stack.push(node)
    }
  }

  // Find body node if present
  let bodyNode: ElementNode | undefined
  for (const n of allNodes) {
    if (n.name === "body") { bodyNode = n; break }
  }

  return { root, allNodes, bodyNode }
}

/**
 * Collect all nodes in a subtree starting from the given root.
 */
function collectSubtree(root: ElementNode): ElementNode[] {
  const collected: ElementNode[] = []
  const stack: ElementNode[] = [root]
  while (stack.length > 0) {
    const current = stack.pop()
    if (!current) continue
    collected.push(current)
    for (let i = current.children.length - 1; i >= 0; i--) {
      stack.push(current.children[i])
    }
  }
  return collected
}

/**
 * Validate that output HTML properly covers all input IDs.
 */
async function validateHTMLStructure(inputFile: string, outputFile: string): Promise<{
  isValid: boolean
  message: string
}> {
  try {
    // Check if files exist
    if (!existsSync(inputFile)) {
      return { isValid: false, message: `Input file does not exist: ${inputFile}` }
    }
    
    if (!existsSync(outputFile)) {
      return { isValid: false, message: `Output file does not exist: ${outputFile}` }
    }
    
    // Read files
    const inputHTML = await readFile(inputFile, "utf-8")
    let outputHTML = await readFile(outputFile, "utf-8")
    
    // Replace <em> with <i> and <strong> with <b> in output
    const replacementsMade: string[] = []
    if (outputHTML.includes("<em>") || outputHTML.includes("</em>")) {
      outputHTML = outputHTML.replace(/<em>/g, "<i>").replace(/<\/em>/g, "</i>")
      replacementsMade.push("<em> → <i>")
    }
    if (outputHTML.includes("<strong>") || outputHTML.includes("</strong>")) {
      outputHTML = outputHTML.replace(/<strong>/g, "<b>").replace(/<\/strong>/g, "</b>")
      replacementsMade.push("<strong> → <b>")
    }
    
    // Write back the modified output if replacements were made
    if (replacementsMade.length > 0) {
      await writeFile(outputFile, outputHTML, "utf-8")
      console.warn(`📝 Auto-replaced tags: ${replacementsMade.join(", ")}`)
    }
    
    // Parse HTML
    const inputTree = parseHTMLTree(inputHTML)
    const outputTree = parseHTMLTree(outputHTML)
  
    // Extract IDs from input
    const idsInInput = new Set<number>()
    for (const element of inputTree.allNodes) {
      if (element.attrs.id) {
        const id = parseInt(element.attrs.id)
        if (isNaN(id)) {
          return { isValid: false, message: `Non-numeric ID found in input: ${element.attrs.id}` }
        }
        idsInInput.add(id)
      }
    }
    
    // Check for disallowed tags in output
    const disallowedTags = new Set<string>()
    // Only validate tags INSIDE <body> (exclude the <body> element itself).
    // If there's no <body>, fall back to validating the full document tree.
    const nodesToCheck: ElementNode[] = outputTree.bodyNode
      ? outputTree.bodyNode.children.flatMap(child => collectSubtree(child))
      : outputTree.allNodes
    for (const element of nodesToCheck) {
      if (!ALLOWED_TAGS.includes(element.name)) {
        disallowedTags.add(element.name)
      }
    }
    
    // Extract IDs from output data-sources attributes (leaf nodes only)
    const idsInOutput = new Set<number>()
    for (const element of outputTree.allNodes) {
      const dataSources = element.attrs["data-sources"]
      if (dataSources) {
        const hasNonInlineChild = element.children.some(child => !INLINE_STYLE_TAGS.has(child.name))
        if (!hasNonInlineChild) {
          try {
            const ids = parseRangeString(dataSources)
            ids.forEach(id => idsInOutput.add(id))
          } catch (error) {
            return { 
              isValid: false, 
              message: `Failed to parse data-sources '${dataSources}': ${error}` 
            }
          }
        }
      }
    }
    
    // Check coverage
    const missingIds = Array.from(idsInInput).filter(id => !idsInOutput.has(id))
    const extraIds = Array.from(idsInOutput).filter(id => !idsInInput.has(id))
    
    if (disallowedTags.size > 0 || missingIds.length > 0 || extraIds.length > 0) {
      const errorMsgs: string[] = []
      
      if (disallowedTags.size > 0) {
        errorMsgs.push(
          `You've used some HTML tags in the ${outputFile} file that are not allowed: ${Array.from(disallowedTags).sort()}.\n` +
          `The allowed tags are: ${ALLOWED_TAGS.join(", ")}.\n` +
          "Fix these tags and keep going! You're doing great!"
        )
      }
      
      if (missingIds.length > 0) {
        let missingMsg = `IDs in the ${inputFile} file not yet covered by leaf nodes in the ${outputFile} file: ${missingIds.sort()}. You'll need to add more leaf nodes (or make sure all existing leaf nodes have data-sources) that cover these ids.`
        if (missingIds.length < 30) {
          missingMsg += (
            "\n\nNote: if any input nodes are empty or contain garbage characters " +
            "you think shouldn't be in the final output, you may attach their ids " +
            "as data-sources to neighboring output nodes to pass the id validation " +
            "check. Please do this sparingly."
          )
        }
        errorMsgs.push(missingMsg)
      }
      
      if (extraIds.length > 0) {
        errorMsgs.push(
          "Output references IDs not present in input (invalid data-sources). " +
          `Extra IDs: ${extraIds.sort()}. ` +
          "Ensure data-sources only contain IDs that exist in the input and split ranges " +
          "to avoid spanning missing IDs."
        )
      }
      
      return { isValid: false, message: errorMsgs.join("; ") }
    }
    
    return { isValid: true, message: "All IDs properly covered" }
    
  } catch (error) {
    return { isValid: false, message: `Validation error: ${error}` }
  }
}

export const ValidateHtmlPlugin: Plugin = async ({ app, client, $ }) => {
  console.log("HTML Validation Plugin initialized!")

  return {
    "tool.execute.after": async ({ tool, sessionID, callID }, output) => {
      // Only validate when HTML files are written
      if (tool !== "write" && tool !== "edit") {
        console.log("Skipping HTML validation for non-write or edit tool", tool)
        return
      }
      
      const inputFile: string = "input.html"
      const outputFile: string = "output.html"
      
      const result = await validateHTMLStructure(inputFile, outputFile)
      
      let validationMessage = `🔍 **HTML Validation**: Checking restructuring of \`${inputFile}\` → \`${outputFile}\`\n\n`
      
      if (result.isValid) {
        // Check for placeholders and length
        const outputHTML = await readFile(outputFile, "utf-8")
        const inputHTML = await readFile(inputFile, "utf-8")
        
        const possiblePlaceholders: string[] = []
        const lines = outputHTML.split("\n")
        for (let i = 0; i < lines.length; i++) {
          if (/placeholder|<!--|\.\.\./.test(lines[i])) {
            possiblePlaceholders.push(`Line ${i + 1}: ${lines[i].trim()}`)
          }
        }
        
        validationMessage += `✅ **Validation Passed!** All IDs from input are present as data-sources in output and all tags are valid.`
        
        if (possiblePlaceholders.length > 0) {
          validationMessage += `\n\n⚠️ **Warning**: Possible placeholders detected. Please review these lines:\n${possiblePlaceholders.map(line => `- ${line}`).join("\n")}`
        }
        
        // If output file is <85% the length of the input file, add warning
        if (outputHTML.length < 0.85 * inputHTML.length) {
          validationMessage += `\n\n⚠️ **Length Warning**: Output is suspiciously short (${Math.round((outputHTML.length / inputHTML.length) * 100)}% of input length). Verify all content is properly represented.`
        }
        
        validationMessage += `\n\n**Next Steps**: Review the output HTML structure and semantic tags. If satisfied, mark the task complete.`
      } else {
        validationMessage += `📝 **Validation In Progress**: ${result.message}\n\nKeep going! You're making progress toward a valid HTML restructuring.`
      }
      
      // Modify the tool's displayed result instead of sending a new chat message
      console.log("Modifying tool's displayed result", validationMessage)
      output.title = "HTML Validation"
      output.output = `${output.output ?? ""}\n\n${validationMessage}`
      output.metadata = { ...(output.metadata || {}), htmlValidation: { inputFile, outputFile, isValid: result.isValid } }
    }
  }
}
